toolchain/check/function.cpp

```diff
@@ -0,0 +1,216 @@
+// Part of the Carbon Language project, under the Apache License v2.0 with LLVM
+// Exceptions. See /LICENSE for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "toolchain/check/function.h"
+
+namespace Carbon::Check {
+
+// Returns true if the function signature has an error, which will have
+// previously been diagnosed.
+static auto FunctionSignatureHasError(Context& context,
+                                      const SemIR::Function& fn) -> bool {
+  if (fn.return_type_id == SemIR::TypeId::Error) {
+    return true;
+  }
+  for (auto param_refs_id : {fn.implicit_param_refs_id, fn.param_refs_id}) {
+    if (param_refs_id != SemIR::InstBlockId::Empty) {
+      for (auto param_id : context.inst_blocks().Get(param_refs_id)) {
+        if (context.insts().Get(param_id).type_id() == SemIR::TypeId::Error) {
+          return true;
+        }
+      }
+    }
+  }
+  return false;
+}
+
+// Returns true if a param agrees. The caller is expected to provide a
+// diagnostic.
+// TODO: Consider moving diagnostics here, particularly to differentiate
+// between type and name mistakes. For now, taking the simpler approach because
+// I also think we may want to refactor params.
+static auto DoesParamAgree(Context& context, SemIR::InstId new_param_ref_id,
+                           SemIR::InstId prev_param_ref_id) -> bool {
+  auto new_param_ref = context.insts().Get(new_param_ref_id);
+  auto prev_param_ref = context.insts().Get(prev_param_ref_id);
+  if (new_param_ref.kind() != prev_param_ref.kind() ||
+      new_param_ref.type_id() != prev_param_ref.type_id()) {
+    return false;
+  }
+
+  if (new_param_ref.Is<SemIR::AddrPattern>()) {
+    new_param_ref =
+        context.insts().Get(new_param_ref.As<SemIR::AddrPattern>().inner_id);
+    prev_param_ref =
+        context.insts().Get(prev_param_ref.As<SemIR::AddrPattern>().inner_id);
+  }
```

Comment:
Hm, I'm torn. I want to suggest using `TryAs` here, but that'd make the two statements in the `if` body less similar. I wonder if it makes sense to get rid of the early `kind` check and instead use

```suggestion
  if (auto new_addr = new_param_ref.TryAs<SemIR::AddrPattern>(),
         prev_addr = prev_param_ref.TryAs<SemIR::AddrPattern>();
       new_addr && prev_addr) {
    new_param_ref = context.insts().Get(new_addr.inner_id);
    prev_param_ref = context.insts().Get(prev_addr.inner_id);
  }
```

What do you think? In part I'm thinking that we'll eventually allow the `unused` wrapper to differ, so the early kind check might not be desirable in general.

(I'm also fine with this as-is, just for your consideration.)

Reply:
I feel weird about this. As discussed, my sense is we need to flatten params more, but I'm not sure what the path we'll take is. Maybe we should just move addr to Param now? But I'm hesitant to make this particular change due to the inconsistency, and partly because I see it as doing two TryAs even if the first fails.